libfwupdplugin/fu-firmware-common.c

```diff
@@ -121,3 +122,158 @@ fu_firmware_strparse_uint32 (const gchar *data)
 	buffer[8] = '\0';
 	return (guint32) g_ascii_strtoull (buffer, NULL, 16);
 }
+
+/**
+ * fu_firmware_strparse_uint4_safe:
+ * @data: destination buffer
+ * @datasz: size of @data, typcally the same as `strlen(data)`
+ * @offset: offset in chars into @data to read
+ * @value: (out) (nullable): parsed value
+ * @error: A #GError or %NULL
+ *
+ * Parses a base 16 number from a string.
+ *
+ * Return value: %TRUE if parsed, %FALSE otherwise
+ *
+ * Since: 1.5.6
+ **/
+gboolean
+fu_firmware_strparse_uint4_safe (const gchar *data,
+				 gsize datasz,
+				 gsize offset,
+				 guint8 *value,
+				 GError **error)
+{
+	gchar buffer[2] = { '\0' };
+	if (!fu_memcpy_safe ((guint8 *) buffer, sizeof(buffer), 0x0,	/* dst */
+			     (const guint8 *) data, datasz, offset,	/* src */
+			     sizeof(buffer) - 1, error))
+		return FALSE;
+	if (value != NULL)
+		*value = (guint8) g_ascii_strtoull (buffer, NULL, 16);
+	return TRUE;
+}
+
+/**
+ * fu_firmware_strparse_uint8_safe:
+ * @data: destination buffer
+ * @datasz: size of @data, typcally the same as `strlen(data)`
+ * @offset: offset in chars into @data to read
+ * @value: (out) (nullable): parsed value
+ * @error: A #GError or %NULL
+ *
+ * Parses a base 16 number from a string.
+ *
+ * Return value: %TRUE if parsed, %FALSE otherwise
+ *
+ * Since: 1.5.6
+ **/
+gboolean
+fu_firmware_strparse_uint8_safe (const gchar *data,
+				 gsize datasz,
+				 gsize offset,
+				 guint8 *value,
+				 GError **error)
+{
+	gchar buffer[3] = { '\0' };
+	if (!fu_memcpy_safe ((guint8 *) buffer, sizeof(buffer), 0x0,	/* dst */
+			     (const guint8 *) data, datasz, offset,	/* src */
+			     sizeof(buffer) - 1, error))
+		return FALSE;
+	if (value != NULL)
+		*value = (guint8) g_ascii_strtoull (buffer, NULL, 16);
+	return TRUE;
+}
+
+/**
+ * fu_firmware_strparse_uint16_safe:
+ * @data: destination buffer
+ * @datasz: size of @data, typcally the same as `strlen(data)`
+ * @offset: offset in chars into @data to read
+ * @value: (out) (nullable): parsed value
+ * @error: A #GError or %NULL
+ *
+ * Parses a base 16 number from a string.
+ *
+ * Return value: %TRUE if parsed, %FALSE otherwise
+ *
+ * Since: 1.5.6
+ **/
+gboolean
+fu_firmware_strparse_uint16_safe (const gchar *data,
+				  gsize datasz,
+				  gsize offset,
+				  guint16 *value,
+				  GError **error)
+{
+	gchar buffer[5] = { '\0' };
+	if (!fu_memcpy_safe ((guint8 *) buffer, sizeof(buffer), 0x0,	/* dst */
+			     (const guint8 *) data, datasz, offset,	/* src */
+			     sizeof(buffer) - 1, error))
+		return FALSE;
+	if (value != NULL)
+		*value = (guint16) g_ascii_strtoull (buffer, NULL, 16);
+	return TRUE;
+}
+
+/**
+ * fu_firmware_strparse_uint24_safe:
+ * @data: destination buffer
+ * @datasz: size of @data, typcally the same as `strlen(data)`
+ * @offset: offset in chars into @data to read
+ * @value: (out) (nullable): parsed value
+ * @error: A #GError or %NULL
+ *
+ * Parses a base 16 number from a string.
```

Comment:
this description should be changed for each function.  It's same right now.

Reply:
IE describe the need for uint24 vs uint16 vs uint8

Reply:
Something like:

     * Parses a base 16 number from a string of 6 characters in length.
     * The returned @value will range from from 0 to 0xffffff.

Reply:
Yes that is exactly what I think we need in these functions

Reply:
Okay, fixed, thanks.